pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -607,8 +625,64 @@ func (c *Controller) getNodePortGatewayServices() []*model.Service {
 	return out
 }
 
-// updateServiceExternalAddr updates ClusterExternalAddresses for ingress gateway service of nodePort type
-func (c *Controller) updateServiceExternalAddr(svcs ...*model.Service) bool {
+func (c *Controller) updateNetworkGateways(svc *model.Service) {
+	c.Lock()
+	defer c.Unlock()
+	svc.Mutex.RLock()
+	defer svc.Mutex.RUnlock()
+	// meshNetworks registryServiceName+fromRegistry
+	if gwPort, ok := c.registryServiceNameGateways[svc.Hostname]; ok {
+		if c.networkGateways[svc.Hostname] == nil {
+			c.networkGateways[svc.Hostname] = map[string][]*model.Gateway{}
+		}
+		c.networkGateways[svc.Hostname][c.networkForRegistry] = c.extractGatewaysFromService(svc, gwPort)
+	}
+	// label based gateways
+	if nw := svc.Attributes.Labels[label.IstioNetwork]; nw != "" {
+		if c.networkGateways[svc.Hostname] == nil {
+			c.networkGateways[svc.Hostname] = map[string][]*model.Gateway{}
+		}
+		gwPort := getGatewayPort(svc)
+		c.networkGateways[svc.Hostname][nw] = c.extractGatewaysFromService(svc, gwPort)
+	}
+}
+
+func getGatewayPort(svc *model.Service) uint32 {
+	if gwPortStr := svc.Attributes.Labels[IstioGatewayPortLabel]; gwPortStr != "" {
+		if gwPort, err := strconv.Atoi(gwPortStr); err != nil {
+			return uint32(gwPort)
+		}
+		log.Warna(fmt.Sprintf("could not parse %q; defaulting to %d",
```

Comment:
nit: put the service name in the log and you can use Warnf instead of Warna+fmt.Sprintf